cpp/src/arrow/c/abi.h

```diff
@@ -106,6 +169,77 @@ struct ArrowArrayStream {
 
 #endif  // ARROW_C_STREAM_INTERFACE
 
+#ifndef ARROW_C_DEVICE_STREAM_INTERFACE
+#define ARROW_C_DEVICE_STREAM_INTERFACE
+
+struct ArrowDeviceArrayStream {
+  // The device that this stream produces data on.
+  // All ArrowDeviceArrays that are produced by this
+  // stream should have the same device_type as set
+  // here. The device_type needs to be provided here
+  // so that consumers can provide the correct type
+  // of stream_ptr when calling get_next.
+  ArrowDeviceType device_type;
+
+  // Callback to get the stream schema
+  // (will be the same for all arrays in the stream).
+  //
+  // Return value: 0 if successful, an `errno`-compatible error code otherwise.
+  //
+  // If successful, the ArrowSchema must be released independently from the stream.
+  int (*get_schema)(struct ArrowDeviceArrayStream*, struct ArrowSchema* out);
+
+  // Callback to get the device id for the next array.
+  // This is necessary so that the proper/correct stream pointer can be provided
+  // to get_next. The parameter provided must not be null.
+  //
+  // Return value: 0 if successful, an `errno`-compatible error code otherwise.
+  //
+  // The next call to `get_next` should provide an ArrowDeviceArray whose
+  // device_id matches what is provided here, and whose device_type is the
+  // same as the device_type member of this stream.
+  int (*get_next_device_id)(struct ArrowDeviceArrayStream*, int* out_device_id);
+
+  // Callback to get the next array
+  // (if no error and the array is released, the stream has ended)
+  //
+  // the provided stream_ptr should be the appropriate stream, or
+  // equivalent object, for the device that the data is allocated on
+  // to indicate where the consumer wants the data to be accessible.
```

Comment:
I'm not sure I understand, why would the consumer be able to choose where the data is accessible? If the data is already allocated (which is certainly a common case), how the producer make it accessible to a new stream?

Reply:
In the CUDA Array Interface, it's the producer that optionally fills in the stream that must be synchronized on to access the data:
https://numba.readthedocs.io/en/stable/cuda/cuda_array_interface.html#synchronization

Besides, there's an interesting statement about stream lifetime here:
https://numba.readthedocs.io/en/stable/cuda/cuda_array_interface.html#streams

Reply:
Stream lifetime is a tough thing to handle for GPUs because everything is async. So after discussing with @kkraus14 we came to the conclusion that the safest API would be for a consumer to specify which stream they wanted to be able to access the data from and the producer would manage any event and stream synchronizing necessary to make it happen. This puts the stream lifetime management into the consumer (where it's easier to manage) rather than the producer.

@kkraus14 you can probably explain this better than I just did.

Reply:
This seems a bit gratuitously inconsistent with the handling of the lifetime of memory buffers (which goes through the release callback, telling the producer when it can safely release them).

It is also gratuitously inconsistent with Numba's CUDA array interface, and I would trust the Numba developers' experience here.

Reply:
I drove a lot of the development on the `__cuda_array_interface__` with @seibert and @sklam and the stream was added later and I haven't seen it used. The challenge is that most frameworks internally manage their streams and don't have the ability to share a stream and control its lifetime nicely.

The idea of a consumer of the interface passing a stream to the producer of the interface is that the producer can guarantee that the data is safe to consume on the passed in stream. A bunch of array library maintainers including folks from Tensorflow, PyTorch, JAX, CuPy, etc. all agreed on these semantics: https://data-apis.org/array-api/latest/API_specification/generated/array_api.array.__dlpack__.html

Reply:
Correct. This is why there isn't a `void*` in the `ArrowDeviceArray` struct itself.